### PR TITLE
smt: abstract multiplication

### PR DESCRIPTION
**Pull request by daejunpark**
Abstract multiplication in a similar way to division and modulo operations.

The actual bitvector multiplication is used when either argument is concrete. Also, when an argument is a power of two, the left-shift is used instead of multiplication.

### CODE REVIEW

File: src/halmos/sevm.py [hunk 1]
```diff
@@ -1348,7 +1356,35 @@ def arith(self, ex: Exec, op: int, w1: Word, w2: Word) -> Word:
...
+            if is_bv_value(w1):
+                i1: int = w1.as_long()
+                if i1 == 0:
+                    return w1
+
+                if i1 == 1:
+                    return w2
+
+                if is_power_of_two(i1):
+                    return w2 << int(math.log(i1, 2))
```
Comment (karmacoma-eth) [inline src/halmos/sevm.py:1371]:
```suggestion
                    return w2 << i1.bit_length() - 1
```

quick check:

~/projects/halmos on code-fastpath wip *21 !2 ?50 ❯ python -m timeit --setup='import math' 'int(math.log(32, 2))'                                                                                                                                                                     
2000000 loops, best of 5: 121 nsec per loop

~/projects/halmos on code-fastpath wip *21 !2 ?50 ❯ python -m timeit 'int(32).bit_length() - 1'                                                                                                                                                                 
5000000 loops, best of 5: 66.1 nsec per loop

File: src/halmos/sevm.py [hunk 2]
```diff
@@ -1348,7 +1356,35 @@ def arith(self, ex: Exec, op: int, w1: Word, w2: Word) -> Word:
             return w1 - w2
 
         if op == EVM.MUL:
-            return w1 * w2
+            if is_bv_value(w1) and is_bv_value(w2):
```
Comment (karmacoma-eth) [inline src/halmos/sevm.py:1359]: can we invoke `is_bv_value(x)` once for each and reuse the results? in the worst case it results in 6 calls to that function, which itself makes more function calls

File: src/halmos/sevm.py [hunk 3]
```diff
@@ -1348,7 +1356,35 @@ def arith(self, ex: Exec, op: int, w1: Word, w2: Word) -> Word:
...
+                return w1 * w2
```
Comment (karmacoma-eth) [inline src/halmos/sevm.py:1360]: just to double check: we're fine here in case of overflow?

Comment (daejunpark) [reply src/halmos/sevm.py:1360]: yes, and it's indeed intended.

File: src/halmos/sevm.py (hunk 1 above)
Comment (daejunpark) [reply src/halmos/sevm.py]: done: https://github.com/a16z/halmos/pull/340/commits/2ee41207c8a111980670ff7f426fc20f1e57b89a

File: src/halmos/sevm.py (hunk 2 above)
Comment (daejunpark) [reply src/halmos/sevm.py]: done: https://github.com/a16z/halmos/pull/340/commits/e8490ee2815662b77474a4bcbf5695ac1c119f4e

**Review by karmacoma-eth — APPROVED** [review]

**Pull request merged**